src/mapreduce.jl

```diff
@@ -0,0 +1,276 @@
+    w = prepare(tmap!, f, dst, srcs, batch_size=batch_size)
+    run!(w)
+    dst
+end
+
+function tmap(f, srcs::AbstractArray...; 
+             batch_size=default_batch_size(length(first(srcs)))
+            )
+    g = Base.Generator(f,srcs...)
+    T = Base.@default_eltype(g)
+    dst = similar(first(srcs), T)
+    tmap!(f, dst, srcs...; batch_size=batch_size)
+end
+
+############################## tmapreduce(like) ##############################
+struct Reduction{O}
+    op::O
+end
+(red::Reduction)(f, srcs...) = Base.mapreduce(f, red.op, srcs...)
+
+struct MapReduceWorkspace{R,F,B,V,OA<:OffsetArray}
+    reduction::R
+    f::F
+    batches::Batches{B}
+    arena_src_views::V
+    batch_reductions::OA
+end
+
+struct NoInit end
+
+function create_reduction(::typeof(tmapreduce), op)
+    Reduction(op)
+end
+
+function prepare(::typeof(tmapreduce), f, op, srcs; init, batch_size::Int)
+    red = create_reduction(tmapreduce, op)
+    w = prepare_mapreduce_like(red, f, srcs, init, batch_size=batch_size)
+    return w
+end
+
+function tmapreduce(f, op, srcs::AbstractArray...;
+                   init=NoInit(),
+                   batch_size= default_batch_size(length(first(srcs)))
+                  )
+    if isempty(first(srcs))
+        if init isa NoInit
+            return Base.mapreduce(f, op, srcs...)
+        else
+            return Base.mapreduce(f, op, srcs..., init=init)
+        end
+    end
+    w = prepare(tmapreduce, f, op, srcs, init=init, batch_size=batch_size)
+    run!(w)
+end
+
+function treduce(op, srcs::AbstractArray...; kw...)
+    tmapreduce(identity, op, srcs...; kw...)
+end
+
+
+for red in SYMBOLS_MAPREDUCE_LIKE
+    tred = tname(red)
+    @eval function $tred end
+
+    @eval function prepare(::typeof($tred), f, srcs; batch_size::Int)
+        base_red = Base.$red
+        prepare_mapreduce_like(base_red, f, srcs, batch_size=batch_size)
+    end
+
+    @eval function $tred(f, src::AbstractArray;
+                        batch_size=default_batch_size(length(src)))
+        isempty(src) && return Base.$red(f, src)
+        srcs = (src,)
+        w = prepare($tred, f, srcs, batch_size=batch_size)
+        run!(w)
+    end
+    @eval $tred(src; kw...) = $tred(identity, src; kw...)
+end
+
+function prepare_mapreduce_like(red, f, srcs, init=NoInit(); batch_size::Int)
+    all_inds  = eachindex(IndexLinear(), srcs...)
+    batches   = Batches(all_inds, batch_size)
+    sample_inds = batches[1]
+
+    arena_src_views = create_arena_src_views(srcs, sample_inds)
+    T = get_return_type(red, f, srcs)
+
+    if (init isa NoInit)
+        batch_reductions = OffsetVector{T}(undef, 1:length(batches))
+    else
+        batch_reductions = OffsetVector{T}(undef, 0:length(batches))
+        batch_reductions[0] = init
+    end
+    MapReduceWorkspace(red, f, batches, arena_src_views, batch_reductions)
+end
+
+@inline function get_return_type(red, f, srcs)
+    Core.Compiler.return_type(red, Tuple{typeof(f), typeof.(srcs)...})
```

Comment:
What about the case where inference fails? The result can be `Union{}`.

Reply:
You are right this is not ideal. However I don't know practical examples of this. I would like to leave it as is and add some heuristics like on master
https://github.com/mohamed82008/KissThreading.jl/blob/master/src/KissThreading.jl#L163
as bug reports arrive. This has the advantage, that it gives test cases and shows that the heuristics are actually useful in practice. Is that an option for you?

Reply:
I recall having this inference failure before so I added that branch. But I don't remember the example. I will give this a test drive in my [TopOpt.jl](https://github.com/mohamed82008/TopOpt.jl), may be I will find it there.

Reply:
Sounds good!